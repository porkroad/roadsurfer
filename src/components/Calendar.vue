<template>
  <div
      class="md:container md:mx-auto"
  >
    <Autocomplete
        @user-select="onSelect"
    />
    <div class="inline-flex">
      <button
          class="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold py-2 px-4 rounded-l"
          @click="lastWeek()"
      >
        Prev
      </button>
      <button
          class="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold py-2 px-4 rounded-r"
          @click="nextWeek()"
      >
        Next
      </button>
    </div>
    <div
        class="grid grid-cols-1 md:grid-cols-7 p-3"
    >
      <div
          v-for="(day, key) in weekData"
          :key="key"
      >
        <CalendarData :date="day"/>
      </div>
    </div>
  </div>
</template>

<script>

import dayjs from 'dayjs';
import Autocomplete from './Autocomplete';
import CalendarData from "./CalendarData";
import {mapActions, mapState} from 'vuex';

export default {
  name: "Calendar",
  components: {
    CalendarData,
    Autocomplete,
  },
  data() {
    return {
      dateToSelect: dayjs(),
    };
  },
  computed: {
    ...mapState('booking', [
      'initData'
    ]),
    weekData() {
      const currentDay = this.dateToSelect;
      const weekData = [];
      if (currentDay.day() !== 0) {
        while (currentDay.subtract(1, 'day').day() !== 0) {
          const date = currentDay.subtract(1, 'day');

          weekData.push({formatted: date.format('MMM ddd D'), timestamp: date.unix() * 1000});
        }
      }

      weekData.push({formatted: currentDay.format('MMM ddd D'), timestamp: currentDay.unix() * 1000});

      let daysToAdd = 0;
      while (weekData.length < 7) {
        const date = currentDay.add(++daysToAdd, 'day');

        weekData.push({formatted: date.format('MMM ddd D'), timestamp: date.unix() * 1000});
      }

      return weekData;
    },
  },
  async mounted() {
    const data = await fetch('https://605c94c36d85de00170da8b4.mockapi.io/stations');
    const initData = await data.json();

    this.setInitData(initData);
  },
  methods: {
    ...mapActions('booking', [
      'setInitData',
      'setSelected'
    ]),
    lastWeek() {
      this.dateToSelect = this.dateToSelect.subtract(1, 'week');
    },
    nextWeek() {
      this.dateToSelect = this.dateToSelect.add(1, 'week');
    },
    onSelect(result) {
      const station = this.initData.find((stations) => stations.name === result);

      this.setSelected(station.id);
    },
  }
}
</script>

<style scoped>

</style>
