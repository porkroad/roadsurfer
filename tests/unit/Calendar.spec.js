import {shallowMount} from '@vue/test-utils'
import Calendar from '@/components/Calendar'

import dayjs from 'dayjs';

describe('Calendar.vue', () => {
    const wrapper = shallowMount(Calendar, {
        methods: {init: jest.fn()}
    });
    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should update week (subtract)', () => {
        wrapper.setData({dateToSelect: dayjs()});

        wrapper.find('.button-last-week').trigger('click');

        const currentDay = dayjs();
        expect(wrapper.vm.dateToSelect.format('Y-m-d')).toEqual(currentDay.subtract(1,'week').format('Y-m-d'));
    });

    it('should update week (add)', () => {
        wrapper.setData({dateToSelect: dayjs()});

        wrapper.find('.button-next-week').trigger('click');

        const currentDay = dayjs();
        expect(wrapper.vm.dateToSelect.format('Y-m-d')).toEqual(currentDay.add(1,'week').format('Y-m-d'));
    });
})
