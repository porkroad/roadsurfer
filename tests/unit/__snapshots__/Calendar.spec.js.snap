// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar.vue should match snapshot 1`] = `
<div class="md:container md:mx-auto">
  <autocomplete-stub></autocomplete-stub>
  <div class="inline-flex"><button class="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold py-2 px-4 rounded-l button-last-week">
      Prev
    </button> <button class="bg-gray-300 hover:bg-gray-400 text-gray-800 font-bold py-2 px-4 rounded-r button-next-week">
      Next
    </button></div>
  <div class="grid grid-cols-1 md:grid-cols-7 p-3">
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
    <div>
      <calendardata-stub date="[object Object]"></calendardata-stub>
    </div>
  </div>
</div>
`;
